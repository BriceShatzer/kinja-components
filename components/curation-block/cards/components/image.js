// @flow

import * as React from 'react';
import type Post from 'kinja-magma/models/Post';
import styled from 'styled-components';
import VideoImageOrThumbnail from '../../../story-cards-stream/figure/video-image-or-thumbnail';
import type { SimpleImageJSON } from 'kinja-magma/models/SimpleImage';
import SimpleImage from 'kinja-magma/models/SimpleImage';
import HeadlineImage from 'kinja-magma/models/HeadlineImage';

type Props = {
	post: Post,
	aboveHeadline?: boolean,
	draggable?: boolean | 'auto',
	sizes?: string,
	noLazy?: boolean,
	customImage?: ?SimpleImageJSON,
	className?: string
}

export const LazyLoadContainer = styled.div`
	padding-bottom: ${9 / 16 * 100}%;
	position: relative;
	height: 0;
`;

export const ImageWrapper = styled.div`
	overflow: hidden;
`;

export default function CardImage(props: Props) {
	const { post, aboveHeadline, draggable, sizes, customImage, noLazy, className } = props;
	const { leftOfHeadline } = post;

	const customHeadlineImage = React.useMemo(() => customImage ? new HeadlineImage({
		...leftOfHeadline,
		id: customImage.id,
		image: SimpleImage.fromJSON(customImage)
	}) : null, [customImage, leftOfHeadline]);

	if (!leftOfHeadline || (!leftOfHeadline && !customHeadlineImage)) {
		return null;
	}

	const showVideoThumbnail = !customImage && (leftOfHeadline.isVideo || post.isVideo);
	return (
		<ImageWrapper className={className}>
			<LazyLoadContainer className={showVideoThumbnail || noLazy ? '' : 'js_lazy-image'}>
				<VideoImageOrThumbnail
					image={customHeadlineImage || leftOfHeadline}
					isAboveHeadline={aboveHeadline || false}
					showVideoThumbnail={showVideoThumbnail}
					croppedImage={true}
					isLeftOfHeadline={!aboveHeadline}
					draggable={draggable}
					sizes={sizes}
					noLazy={noLazy}
				/>
			</LazyLoadContainer>
		</ImageWrapper>
	);
}