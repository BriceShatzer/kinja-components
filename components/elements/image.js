/* @flow */

import * as React from 'react';
import imageUrl from 'kinja-images/imageUrl';
import { omit } from 'lodash';

import { imageFormatFromString, type ImageFormat } from 'postbody/Image';
import ImageSizes, { type ImageSizesT, type ImageTransformsT } from 'kinja-images/config/image-sizes';
import OnScreen from '../hoc/on-screen';
import type { OnScreenInjectedProps } from '../hoc/on-screen';
import {
	withPlatform,
	type PlatformInjectedProps,
	withFeatures,
	type FeaturesInjectedProps
} from '../hoc/context';

export const placeholderSrc = 'data:image/gif;base64,R0lGODlhAQABAAAAACH5BAEKAAEALAAAAAABAAEAAAICTAEAOw==';

const MEDIA_SIZES: Array<ImageTransformsT> = ['KinjaCenteredMediumAuto', 'KinjaCenteredLargeAuto',
	'CenteredWideExtraLargeAuto', 'CenteredWideExtraBigAuto', 'CenteredWideSuperLarge', 'AvatarSmallAuto'];
export const UNCROPPED_MEDIA_SIZES: Array<ImageTransformsT> =
	['AvatarSmallAuto', 'KinjaUncroppedMedium', 'KinjaUncroppedLargeAuto', 'UncroppedWideExtraLargeAuto', 'WideSuperLargeAuto'];
export const MIN_IMAGE_WIDTH = 320;
export const MAX_POST_IMAGE_WIDTH = typeof window !== 'undefined' ? Math.min(636, window.innerWidth) : 636;
export const MAX_LEFT_OF_HEADLINE_IMAGE_WIDTH = 450;
export const MAX_SIDEBAR_IMAGE_WIDTH = 450;

type LazyImageProps = {
	// Image alt attribute
	alt?: string,
	// Image aria-label attribute
	ariaLabel?: string,
	// Whether or not to disable animation & fetch as a jpg
	noAnimate?: boolean,
	// Whether to allow lazy loading via OnScreen props
	noLazy?: boolean,
	width?: number,
	height?: number
};

export type LazyResponsiveImageProps = LazyImageProps &
	PlatformInjectedProps &
	FeaturesInjectedProps &
	OnScreenInjectedProps & {
		// Optional function prop to trigger when image is loaded
		onLoad?: () => void,
		// Optional function prop to trigger when video is loaded
		onCanPlay?: () => void,
		// The image format of the image
		format?: ImageFormat,
		// The id of the image
		id: string,
		//  Whether to show cropped image
		croppedImage?: boolean,
		// Prop for the image to tell the browser which image to choose from the srcset
		// https://developer.mozilla.org/en-US/docs/Learn/HTML/Multimedia_and_embedding/Responsive_images#Resolution_switching_Different_sizes
		sizes?: string,
		transform?: string,
		animated?: boolean,
		draggable?: boolean | 'auto',
		kinjaMeta?: {
			features?: ?{ [name: string]: boolean }
		}
	};

/*
 * A lazy-loading image element that takes a source url
 */
const Image = ({
	id,
	format = 'jpg',
	transform = 'KinjaCenteredExtraLarge',
	onImageLoaded,
	...props
}: LazyImageProps &
	OnScreenInjectedProps & {
		// Cloudinary ID (if applicable)
		id?: string,
		format?: ImageFormat,
		transform?: string,
		props?: mixed,
		onImageLoaded?: () => void
	}) => {
	if (id) {
		return <img
			src={imageUrl(id, transform, imageFormatFromString(format))}
			{...omit(props, 'isVisible')}
			onLoad={onImageLoaded} />;
	} else {
		return null;
	}
};

export const calcSrcSet = (
	imageSizes: ImageSizesT,
	transforms: Array<ImageTransformsT>,
	id: string,
	ensuredFormat?: ImageFormat,
	noAnimate?: boolean,
	imageWidth?: number,
	returnFullSet?: boolean): Array<string> => {

	return transforms.reduce((acc, transform) => {
		const transformWidth = parseInt(imageSizes[transform].width, 10);
		// 1.3 is to allow some range
		if (!transformWidth || !returnFullSet && imageWidth && transformWidth / 1.2 > imageWidth) {
			return acc;
		}
		const sourceUrl = imageUrl(id, transform, noAnimate ? 'jpg' : ensuredFormat);
		return acc.concat([`${sourceUrl} ${transformWidth}w`]);
	}, []);
};

function ResponsiveImage(props: LazyResponsiveImageProps) {
	const {
		id,
		format = 'jpg',
		width,
		height,
		noLazy,
		alt,
		sizes,
		transform,
		noAnimate,
		ariaLabel,
		croppedImage = false,
		platform,
		draggable = 'auto'
	} = props;
	const ensuredFormat = imageFormatFromString(format);
	const posterTransform = transform || 'KinjaCenteredLargeAutoFrozen';
	const animatedTransform = transform || 'KinjaCenteredExtraLarge';
	const animated = ensuredFormat && ['gif'].includes(ensuredFormat);
	const transforms = croppedImage ? MEDIA_SIZES : UNCROPPED_MEDIA_SIZES;
	const imageWidth = width && width < MIN_IMAGE_WIDTH ? MIN_IMAGE_WIDTH : width;
	const srcSet = calcSrcSet(ImageSizes, transforms, id, ensuredFormat, noAnimate, imageWidth, Boolean(sizes));

	if (platform === 'amp') {
		return (
			<amp-img
				srcSet={animated ? srcSet[0] : srcSet.join(', ')}
				width={width}
				height={height}
				layout='responsive'
			/>
		);
	}

	const webmSrc = id && imageUrl(id, animatedTransform, 'webm');
	const mp4Src = id && imageUrl(id, animatedTransform, 'mp4');
	const posterSrc = id && imageUrl(id, posterTransform, 'jpg');
	const videoAttributes = {
		loop: true,
		autoPlay: true,
		muted: true,
		playsInline: true
	};
	const attributesForLightbox = {
		'data-chomp-id': id,
		'data-format': format,
		'data-alt': alt,
		'data-anim-src': animated ? posterSrc : ''
	};

	// not lazy loaded image and animated video
	if (noLazy) {
		if (animated) {
			return (
				<video
					poster={posterSrc}
					{...videoAttributes}
					{...attributesForLightbox}
				>
					{mp4Src && <source type='video/mp4' src={mp4Src} />}
					{webmSrc && <source type='video/webm' src={webmSrc} />}
				</video>
			);
		} else {
			return (
				<img
					src={placeholderSrc}
					srcSet={srcSet.join(', ')}
					sizes={sizes}
					alt={alt}
					aria-label={ariaLabel}
					draggable={draggable}
					{...attributesForLightbox}
				/>
			);
		}
	}

	// lazy loaded image and animated video
	if (animated) {
		return (
			<video
				poster={placeholderSrc}
				data-webmsrc={webmSrc}
				data-mp4src={mp4Src}
				data-postersrc={posterSrc}
				{...videoAttributes}
				{...attributesForLightbox}
			/>
		);
	} else {
		return (
			<img
				src={placeholderSrc}
				alt={alt}
				data-srcset={srcSet.join(', ')}
				sizes={sizes}
				aria-label={ariaLabel}
				draggable={draggable}
				{...attributesForLightbox}
			/>
		);
	}
}

const LazyImage = OnScreen(Image, { offset: 100 });
const LazyResponsiveImage = withPlatform(withFeatures(OnScreen(
	ResponsiveImage, { offset: 200, once: true, partialVisibility: true }
)));

export default LazyImage;
export { ResponsiveImage, Image, LazyResponsiveImage };
