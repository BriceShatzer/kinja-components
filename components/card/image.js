/* @flow */

import * as React from 'react';
import { LazyResponsiveImage } from '../elements/image';
import type { ImageFormat } from 'postbody/Image';
import Caption from '../elements/caption';
import Attribution from './attribution';
import Permalink from './permalink';

import type { ImageAttributionType } from 'postbody/Image';

type Props = {
	id?: string,
	format?: ImageFormat,
	width?: number,
	sizes?: string,
	attributionData?: Array<ImageAttributionType>,
	caption?: string,
	isEditor?: boolean,
	captionOnBlur?: string => mixed,
	captionOnClick?: string => mixed,
	croppedImage?: boolean,
	relative?: boolean,
	noAnimate?: boolean,
	withAttribution?: boolean,
	withSrcset?: boolean,
	href?: string,
	isEditMode?: boolean,
	ga?: string,
	gaMobile?: string,
	noLink?: boolean
};

const Image = (props: Props) => {
	const {
		id,
		format,
		width,
		sizes,
		attributionData,
		caption,
		captionOnBlur,
		captionOnClick,
		croppedImage,
		relative = false,
		isEditor = false,
		noAnimate,
		withAttribution,
		href,
		title,
		isEditMode,
		ga,
		gaMobile,
		openInNewTab,
		noLink
	} = props || {};

	const target = openInNewTab ? '_blank' : undefined;

	const responsiveOrRegularImage = <LazyResponsiveImage
		className="image__img"
		key={`img-${id}`}
		id={id}
		format={format}
		width={width}
		sizes={sizes}
		croppedImage={croppedImage}
		relative={relative}
		noAnimate={noAnimate}
		noLazy
	/>;

	const inner = (
		<React.Fragment>
			{responsiveOrRegularImage}
			{caption ? (
				<Caption isEditable={isEditor} dangerouslySetInnerHTML={{ __html: caption }} onBlur={captionOnBlur} onClick={captionOnClick} />
			) : null}
			{withAttribution ? <Attribution isEditor={isEditor} attributions={attributionData} /> : null}
		</React.Fragment>
	);

	return (
		<div className="image-container-wrapper">
			<div className="image-container">
				{noLink ? inner : <Permalink href={href} target={target} title={title} unlink={isEditMode} ga={ga} gaMobile={gaMobile}>{inner}</Permalink>}
			</div>
		</div>
	);
};

export default Image;
