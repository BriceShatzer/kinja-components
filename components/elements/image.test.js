import * as React from 'react';
import { shallow } from 'enzyme';
import { ResponsiveImage, calcSrcSet } from './image';
import ImageSizes from 'kinja-images/config/image-sizes';

const MOCKED_IMAGE = {
	id: 'test_image_id',
	format: 'jpg',
	width: 800,
	height: 443,
	alignment: 'Center',
	caption: [],
	attribution: [],
	altText: undefined,
	relative: true,
	alt: 'yo, this is alt'
};

describe('Image elements', () => {
	it('should render a responsive image', () => {
		const wrapper = shallow(
			<ResponsiveImage
				noLazy
				id={MOCKED_IMAGE.id}
				width={MOCKED_IMAGE.width}
			/>
		);

		expect(wrapper).toMatchSnapshot();
	});

	describe('srcSet calc', () => {
		test('should return array of empty strings when image width is too small', () => {
			const srcSet = calcSrcSet(ImageSizes, ['KinjaCenteredMediumAuto', 'KinjaCenteredLargeAuto',
				'CenteredWideExtraBigAuto', 'CenteredWideSuperLarge', 'AvatarSmallAuto'],
			'test_image_id', true, 'jpg', 60);

			expect(srcSet).toEqual([]);
		});

		test('should return a correct SrcSet for a post width image', () => {
			const srcSetMock =
				['https://i.kinja-img.com/gawker-media/image/upload/c_fill,f_auto,fl_progressive,g_center,h_180,q_80,w_320/test_image_id.jpg 320w',
					'https://i.kinja-img.com/gawker-media/image/upload/c_fill,f_auto,fl_progressive,g_center,h_264,q_80,w_470/test_image_id.jpg 470w',
					'https://i.kinja-img.com/gawker-media/image/upload/c_fill,f_auto,fl_progressive,g_center,h_80,q_80,w_80/test_image_id.jpg 80w'];
			const srcSet = calcSrcSet(ImageSizes, ['KinjaCenteredMediumAuto', 'KinjaCenteredLargeAuto',
				'CenteredWideExtraBigAuto', 'CenteredWideSuperLarge', 'AvatarSmallAuto'],
			'test_image_id', true, 'jpg', 636);

			expect(srcSet).toEqual(srcSetMock);
		});
	});

	describe('data attributes are present', () => {
		let image;

		beforeAll(() => {
			image = shallow(<ResponsiveImage {...MOCKED_IMAGE} />).find('img');
		});

		test('data-srcset', () => {
			expect(image.prop('data-srcset')).toMatchSnapshot();
		});

		describe('when animated', () => {
			let image;
			const props = Object.assign({}, MOCKED_IMAGE, { format: 'gif' });

			beforeAll(() => {
				image = shallow(<ResponsiveImage {...props} />).find('video');
			});


			test('data-postersrc should be the default transform as jpg', () => {
				expect(image.prop('data-postersrc'))
					.toEqual(
						expect.stringContaining('test_image_id.jpg')
					);
			});

			test('data-webmsrc', () => {
				expect(image.prop('data-webmsrc'))
					.toEqual(
						expect.stringContaining('test_image_id.webm')
					);
			});

			test('data-mp4src', () => {
				expect(image.prop('data-mp4src'))
					.toEqual(
						expect.stringContaining('test_image_id.mp4')
					);
			});
		});
	});
});
