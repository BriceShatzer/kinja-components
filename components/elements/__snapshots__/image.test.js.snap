// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Image elements data attributes are present data-srcset 1`] = `"https://i.kinja-img.com/gawker-media/image/upload/c_fill,f_auto,fl_progressive,g_center,h_80,q_80,w_80/test_image_id.jpg 80w, https://i.kinja-img.com/gawker-media/image/upload/c_fit,fl_progressive,q_80,w_320/test_image_id.jpg 320w, https://i.kinja-img.com/gawker-media/image/upload/c_fit,f_auto,fl_progressive,q_80,w_470/test_image_id.jpg 470w, https://i.kinja-img.com/gawker-media/image/upload/c_scale,f_auto,fl_progressive,q_80,w_800/test_image_id.jpg 800w"`;

exports[`Image elements should render a responsive image 1`] = `
<img
  data-anim-src=""
  data-chomp-id="test_image_id"
  data-format="jpg"
  draggable="auto"
  src="data:image/gif;base64,R0lGODlhAQABAAAAACH5BAEKAAEALAAAAAABAAEAAAICTAEAOw=="
  srcSet="https://i.kinja-img.com/gawker-media/image/upload/c_fill,f_auto,fl_progressive,g_center,h_80,q_80,w_80/test_image_id.jpg 80w, https://i.kinja-img.com/gawker-media/image/upload/c_fit,fl_progressive,q_80,w_320/test_image_id.jpg 320w, https://i.kinja-img.com/gawker-media/image/upload/c_fit,f_auto,fl_progressive,q_80,w_470/test_image_id.jpg 470w, https://i.kinja-img.com/gawker-media/image/upload/c_scale,f_auto,fl_progressive,q_80,w_800/test_image_id.jpg 800w"
/>
`;
